### PR TITLE
ENH: improve stats.binned_statistic_dd performance

### PR DESCRIPTION
**Pull request by AtsushiSakai**
#### Reference issue
Related #11468 

#### What does this implement/fix?
In #11757, we improved `binned_statistic_dd` performance for `std` to avoid unnecessary np.std call.

I achieved more better performances only for `max`, `min`, `median`, `std` statistic when I improved binned data search.
I changed to create a binned data hashmap before calculating statistics because searching binned values 

`values[vv, binnumbers == i]`

looks bottleneck.
 
Also, I improved the `test_dd_multi_values` test to test all statistics, and I added benchmarks to evaluate performance in all statistics.

#### Additional information

These are the benchmark result

Before:
SS 2020-06-21 at 20 52 32

After:
SS 2020-06-21 at 23 28 53

`max` and `min` improved about 50%, `std` improved about 70%, and `median`  improved about 15%.

### CODE REVIEW

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py:659]: Were trying to use a `dict` ? If so, use `dict()` IIRC is b/c parsers conflate empty dicts w/sets.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py:647]: the unique calculation is also done outside the loop, isn't it not needed here then?

File: scipy/stats/tests/test_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/tests/test_binned_statistic.py:441]: why the `np.std` is here,  using as an example of passing in `callable`, or is there any other reason?

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py:588]: the signature/naming here is weird, `binnumbers` and `unique_bin_numbers` shouldn't both be necessary information, is there a reason I'm missing?

File: benchmarks/benchmarks/stats.py
Comment (rlucas7) [inline benchmarks/benchmarks/stats.py:168]: for these b/c everything is the same except the `statistic` argument, would it make sense to use:
https://asv.readthedocs.io/en/stable/writing_benchmarks.html#parametrized-benchmarks
instead? It might be cleaner code but not sure about the actual behavior of the benchmarking, I'd guess the behavior is the same.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py:654]: same comment as above about signature weirdness w/ `binnumbers` and `unique_bin_numbers`

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py:651]: I think there is but asking so that you can confirm, is there ***always*** a `callable()` check already done in outer scope?

**Review by rlucas7 — CHANGES_REQUESTED** [review]
@AtsushiSakai looks like a good approach to ameliorate the runtime slowdown, thanks for working on this one. 

Needs some changes though before we can merge.

File: scipy/stats/_binned_statistic.py
Comment (AtsushiSakai) [reply scipy/stats/_binned_statistic.py]: Oh. that's right. I fixed it.

File: scipy/stats/_binned_statistic.py
Comment (AtsushiSakai) [reply scipy/stats/_binned_statistic.py:637]: What do you mean? the unique calculation is only done here.

File: scipy/stats/tests/test_binned_statistic.py
Comment (AtsushiSakai) [reply scipy/stats/tests/test_binned_statistic.py:441]: Because, I think we should test all if-statement paths including `callable`. If we use `callable', the test should pass.

File: scipy/stats/_binned_statistic.py
Comment (AtsushiSakai) [reply scipy/stats/_binned_statistic.py]: I want to reduce unnecessary calculation as much as possible, so I just call `unique` just one time at outer loop, and reuse it in the `Vdim` for loop. That's why I pass `binnumbers` and `unique_bin_numbers` to the` _create_bin_map` function. I

File: benchmarks/benchmarks/stats.py
Comment (AtsushiSakai) [reply benchmarks/benchmarks/stats.py]: Oh, very good idea. I changed it. Thanks.

File: scipy/stats/_binned_statistic.py
Comment (AtsushiSakai) [reply scipy/stats/_binned_statistic.py:658]: Same comment above. Should we only pass `binnumbers', and the `unique` function should be called inside function? But It calls 'unique' functions multiple times...

File: scipy/stats/_binned_statistic.py
Comment (AtsushiSakai) [reply scipy/stats/_binned_statistic.py:641]: Do you mean L606-L613 check in the latest code? Yes, it already done when `callable` passed as a statistic.

**Comment by AtsushiSakai** [comment]
This is latest benchmark result for later reference.

SS 2020-06-23 at 21 01 36

SS 2020-06-23 at 21 01 55

This result indicates "reuse_bin" does not help "std", "median", I think because `np.std` and `np.median` are bottleneck.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [reply scipy/stats/_binned_statistic.py:658]: The thing that I'm unclear about (or was until re-reading tonight) is why the signature needs both `binnumbers` and `unique_bin_numbers`. The point is to construct a hashmap from unique binnumber to binnumbers, I still think the naming of arguments is/are confusing though.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py:661]: Ah this part helps clarify, the key is the bin number (unique) and the value is a list (or array-like) of binned data. Maybe the reason why I'm confused is because of conflating the word `map` with `hashmap`, can we maybe call the function `_create_bin_hashmap`? 

Not sure if this is just me or if it is (or will be) genuinely confusing to others too.

File: scipy/stats/tests/test_binned_statistic.py
Comment (rlucas7) [reply scipy/stats/tests/test_binned_statistic.py:441]: Ok that aligns with what I was thinking. If you go to the docstring here: 
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.binned_statistic_dd.html
You can see it says:
`‘count’ : compute the count of points within each bin. This is identical to an unweighted histogram. values array is not referenced.`
I'm not sure if we have a test on that here, if not can we add one? Presumably the way to do this would be to add something that would normally throw errors (e.g. contains Nan and/or inf but does not throw error in this case b/c is not referenced). Of course if you've (or anyone else reading) has got a better idea, please let me know. If it makes more sense for the test I'm proposing to be in a separate class that's ok too.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [reply scipy/stats/_binned_statistic.py:641]: I see it as line 520 in master currently, 

https://github.com/scipy/scipy/blob/master/scipy/stats/_binned_statistic.py#L520

I think you are referring to the block of the if/elif/.../elif where it falls to the `callable` after checking keywords.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [reply scipy/stats/_binned_statistic.py:637]: I'm not sure why I thought that yesterday, sometimes the view in the github page is weird, maybe that is what happened. From what I see now I agree with your statement.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py:608]: Just a side comment, looking at this now, it makes it faster ***and*** more readable (IMO). The PR removes some lines of code that are replicated boilerplate.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py:644]: why is it needed to have a ***separate*** function for the `_calc_binned_std`? is this purely b/c of the special logic with the bins have <2 entries? If so what is the performance reduction by not checking for <2 entries in bin? I'd guess that the repeated searching/binning is/was the larger performance reduction cause. 

Thoughts?

File: scipy/stats/_binned_statistic.py
Comment (AtsushiSakai) [reply scipy/stats/_binned_statistic.py:654]: I agree. I changed it to `_create_bin_hashmap`.

File: scipy/stats/_binned_statistic.py
Comment (AtsushiSakai) [reply scipy/stats/_binned_statistic.py]: I took a benchmark, there is no big difference performances, so I merged these functions. PTAL

File: benchmarks/benchmarks/stats.py
Comment (rlucas7) [inline benchmarks/benchmarks/stats.py:155]: good call to distinguish from other 2

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py:652]: Sorry to waffle here but can we remove the hashmap in the ***name*** and change the  docstring to refer say something like "Create hashmap of bin ids to values in bins"? 

Reason being: Hungarian notation (types in names) isn't a good idea for duck typed languages. I realize this is a private function so this is primarily our future selves we are being kind to by choosing descriptive names. 

Pluse if we decide in the future to change the implementation from hashmap to some other data structure, the name would need to change if it contains hashmap. If we remove the implementation becomes something independent of the name. Hope that helps clarify.

**Review by rlucas7 — CHANGES_REQUESTED** [review]
@AtsushiSakai one minor change left and should be good to merge. If you can, it would be great to rebase + squash all commits to a single commit to keep a clean git history on mainline, but not strictly required to merge, just the 1 small name change I request in this latest review.

File: scipy/stats/_binned_statistic.py
Comment (AtsushiSakai) [reply scipy/stats/_binned_statistic.py]: OK. I changed it to `_create_binned_data`.

**Comment by AtsushiSakai** [comment]
Thank you for reviewing. I squashed it.

**Review by rlucas7 — APPROVED** [review]
LGTM. Thanks @AtsushiSakai , marking as approved for now. I'll merge the PR Sunday (July 5th) if no further comments.

**Pull request merged**
